### PR TITLE
subsys: fb: cfb: support drawing to any coordinates 

### PR DESCRIPTION
**Pull request by soburi**
This PR intends to use cfb like as a graphics oriented devices. It will be fundamental for implements simple graphics rendering.

# API improvement:

- Allow printing and inverting at any coordinates.

  Currently, the CFB API can only draw at tile bounds coordinates (where the y coordinate is a multiple of 8).
  Implements dot-by-dot font rendering to make cfb_print() able to draw text even at a coordinate that is not on tile boundaries.
  For API consistency, change cfb_invert_area() able to invert even at a coordinate not on tile boundaries.

- Add `cfb_draw_text()` API

  Added `cfb_draw_text()` that does not wrap when specifying the position in dots, as it may be inconvenient if there is wrapping

# Shell Improvements:

- Add command `draw text`

  Add command to run the new API `cfb_draw_text()`.
  This command is similar to `print`.
  The position specify with (x,y) not column-row and doesn't clear buffer is differences.

- Add support for running cfb_invert_area().
  
  Extend the invert command to run cfb_invert_area when given four arguments.
  Make a change to invert the screen immediately in case of run invert without arguments.

### CODE REVIEW

**Comment by MaureenHelm** [comment]
@jfischer-no will you please have a look?

File: subsys/fb/cfb.c [hunk 1]
```diff
@@ -105,24 +105,39 @@ static uint8_t draw_char_vtmono(const struct char_framebuffer *fb,
...
-			uint8_t byte;
-
-			if ((fb_y + x + g_x) >= fb->size) {
-				return 0;
+		int16_t fb_x = x + g_x;
+		uint8_t byte = get_glyph_byte(glyph_ptr, fptr, g_x, 0);
+
+		for (size_t g_y = 0; g_y < fptr->height; g_y++) {
+			int16_t fb_y = y + g_y;
+			size_t fb_index = (fb_y / 8U) * fb->x_res + fb_x;
```
Comment (jfischer-no) [inline subsys/fb/cfb.c:113]: Would be nice to have short and concise comments on few points, otherwise no one can understand what happens at first glance.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: subsys/fb/cfb.c [hunk 2]
```diff
@@ -87,8 +87,8 @@ static inline uint8_t get_glyph_byte(uint8_t *glyph_ptr, const struct cfb_font *
  * Draw the monochrome character in the monochrome tiled framebuffer,
  * a byte is interpreted as 8 pixels ordered vertically among each other.
  */
-static uint8_t draw_char_vtmono(const struct char_framebuffer *fb,
-			     char c, uint16_t x, uint16_t y)
+static uint8_t draw_char_vtmono(const struct char_framebuffer *fb, char c, uint16_t x, uint16_t y,
```
[the remark this answers is not in the document]
Comment (soburi) [reply subsys/fb/cfb.c]: Thank you for reviewing.

Align the line start.

File: subsys/fb/cfb.c (hunk 1 above)
Comment (soburi) [reply subsys/fb/cfb.c]: Add explaining the purpose of each in the process branching.

File: subsys/fb/cfb.c [hunk 3]
```diff
@@ -105,24 +105,39 @@ static uint8_t draw_char_vtmono(const struct char_framebuffer *fb,
...
-			fb->buf[fb_y + x + g_x] = byte;
+			if ((y % 8) == 0) {
+				if (draw_bg) {
+					fb->buf[fb_index] = byte;
+				} else {
+					fb->buf[fb_index] |= byte;
+				}
+				g_y += 7;
+			} else {
+				if (byte & BIT(g_y % 8)) {
```
[the remark this answers is not in the document]
Comment (soburi) [reply subsys/fb/cfb.c]: It make a sense.

I rewrite this code to run less times loop.

**Review by jfischer-no — DISMISSED** [review]

File: include/zephyr/display/cfb.h [hunk 4]
```diff
@@ -89,6 +89,20 @@ struct cfb_font {
...
+ * and not line wrapping.
+ *
+ * @param dev Pointer to device structure for driver instance
+ * @param str String to print
+ * @param x Position in X direction of the beginning of the string
+ * @param y Position in Y direction of the beginning of the string
+ *
+ * @return 0 on success, negative value otherwise
+ */
+int cfb_draw_text(const struct device *dev, char *str, int16_t x, int16_t y);
```
Comment (5frank) [inline include/zephyr/display/cfb.h:104]: Would be nice with `const char *str`!?  `cfb_print` should probably also have const parameter IMO.

**Comment by teburd** [comment]
@jfischer-no @5frank please look again

**Review by jfischer-no — APPROVED** [review]

**Comment by soburi** [comment]
@5frank 
ping

**Comment by MaureenHelm** [comment]
@fabiobaltieri please take a look

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**